Framework/DataObjects/inc/MantidDataObjects/ScanningWorkspaceBuilder.h

```diff
@@ -0,0 +1,112 @@
+  scanning workspace (a workspace with moving detectors), where all the
+  information about the scan is known in advance. The constructor takes the
+  arguments for the basic construction, then checks are made for consistency as
+  other information about the scanning workspace is set.
+
+  Things that must be set for successful building:
+   - Number of detectors, number of time indexes and number of bins (set via the
+  constructor)
+   - The instrument set via setInstrument
+   - The time ranges set via setTimeRanges
+
+  Some helper methods exist for specific cases, such as the whole instrument
+  rotating around the sample.
+
+  One current limitation to note here, that is not a general restriction within
+  Mantid, is that every detector must have the same set of time indexes.
+
+  Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_DATAOBJECTS_DLL ScanningWorkspaceBuilder {
+public:
+  enum class IndexingType { DEFAULT, TIME_ORIENTED, DETECTOR_ORIENTED };
+
+  ScanningWorkspaceBuilder(const size_t nDetectors, const size_t nTimeIndexes,
+                           const size_t nBins);
+
+  void setHistogram(const HistogramData::Histogram &histogram);
+
+  void setInstrument(
+      const boost::shared_ptr<const Geometry::Instrument> &instrument);
+  void setTimeRanges(const std::vector<
+      std::pair<Kernel::DateAndTime, Kernel::DateAndTime>> &timeRanges);
+  void setTimeRanges(const Kernel::DateAndTime &startTime,
+                     const std::vector<double> &durations);
+  void setPositions(const std::vector<std::vector<Kernel::V3D>> &positions);
+  void setRotations(const std::vector<std::vector<Kernel::Quat>> &rotations);
+  void setInstrumentAngles(const std::vector<double> &instrumentAngles);
+
+  void setIndexingType(const IndexingType indexingType);
+
+  API::MatrixWorkspace_sptr buildWorkspace();
+
+private:
+  size_t m_nDetectors;
+  size_t m_nTimeIndexes;
+  size_t m_nBins;
+
+  HistogramData::Histogram m_histogram;
+
+  boost::shared_ptr<const Geometry::Instrument> m_instrument;
```

Comment:
Just an idea - for member variables that are mandatory, one way you might wish to avoid logical misuse errors is to wrap them as `boost::optional`. That way there is no danger of blindly accepting default initialised values if setters have not been called. You can always check that these have been initialised prior to use via `boost::optional::operator*()`